### PR TITLE
James-3096 handle mailbox capabilities

### PR DESCRIPTION
**Pull request by remk**
The commit `JAMES-3096 mailboxes capabilities add unit tests` should probably be removed, but else it can be reviewed

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala [hunk 1]
```diff
@@ -101,21 +101,32 @@ class JMAPApiRoutes (val authenticator: Authenticator,
...
+            .header(CONTENT_TYPE, JSON_CONTENT_TYPE)
+            .sendString(
+              SMono.fromCallable(() =>
+                serializer.serialize(ResponseObject(ResponseObject.SESSION_STATE, invocations)).toString),
+              StandardCharsets.UTF_8
+            ).`then`())
+        )
+      case false =>
+        SMono.justOrEmpty(methodsByName.get(invocation.methodName))
+          .flatMap(method => SMono.fromPublisher(method.process(invocation, mailboxSession)))
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:122]: Why do we keep trying processing the method if it exists in the case the capabilities are missing from the request?

**Comment by remk** [comment]
> One remark:
> 
> Your tests validates if the following capabilities are supported or not.
> 
> Maybe the ticket description was not clear but what was expected is that given specified capabilities, the properties of the returned Mailbox object gets filtered.

What we understood was that there was two level of capability validation.
The first one for validating that the client use only supported capabilities.
A second one to eventually change the response depending of some capabilites.

For the second point René hinted that there was no applicable capabilites for get mailbox. 
We do not find any by a quick look to the RFC.
Do you know if there is some ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala (hunk 1 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala:122]: yeah we just commited this wip commit at the end of the day :)

**Comment by chibenwa** [comment]
> Do you know if there is some ?

Yes. Read https://github.com/linagora/james-project/issues/3502 .

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+               |        "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+               |        "ids": null
+               |      },
+               |      "c1"]]
+               |}""".stripMargin)
+    .when()
+      .post()
+    .`then`
+      .statusCode(HttpStatus.SC_OK)
+      .contentType(ContentType.JSON)
```
Comment (nvduc91) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala:68]: without assert how you can prove it successful?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala [hunk 3]
```diff
@@ -43,34 +45,66 @@ class MailboxGetMethod @Inject() (serializer: Serializer,
...
   override val methodName: MethodName = MethodName("Mailbox/get")
 
-  override def process(invocation: Invocation, mailboxSession: MailboxSession): Publisher[Invocation] = {
+  object MailboxGetResults {
+    def found(mailbox: Mailbox): MailboxGetResults = MailboxGetResults(Set(mailbox), NotFound(Set.empty))
+    def notFound(mailboxId: MailboxId): MailboxGetResults = MailboxGetResults(Set.empty, NotFound(Set(mailboxId)))
+  }
+
+  case class MailboxGetResults(mailboxes: Set[Mailbox], notFound: NotFound) {
+    def merge(other: MailboxGetResults): MailboxGetResults = MailboxGetResults(this.mailboxes ++ other.mailboxes, this.notFound.merge(other.notFound))
```
Comment (nvduc91) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala:54]: what is that mean? i just ask

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    assertThatJson(response).isEqualTo(
+      """
+        |{
+        |  "type": "urn:ietf:params:jmap:error:unknownCapability",
+        |  "status": 400,
+        |  "detail": "The Request object used capability
+        |    'https://example.com/apis/foobar', which is not supported
+        |    by this server."
+        |}""".stripMargin)
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala:106]: Is this test passing? I don't think I saw it treated in the code

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .build
+  }
+
+  @Test
+  def getMailboxesWithSupportedCapabilitiesShouldBeSuccessful(): Unit = {
+    `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(s"""{
```
Comment (blackheaven) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala:52]: why `s`?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala [hunk 6]
```diff
@@ -201,7 +201,8 @@ class Serializer @Inject() (mailboxIdFactory: MailboxId.Factory) {
       })
     }
 
-  private implicit val mailboxWrites: Writes[Mailbox] = Json.writes[Mailbox]
+  implicit def mailboxWrites(propertiesToHide: Set[String]): Writes[Mailbox] = Json.writes[Mailbox]
+    .transform( (o: JsObject) => JsObject(o.fields.filterNot(entry => propertiesToHide.contains(entry._1))))
```
Comment (blackheaven) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:205]: extra space

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala [hunk 7]
```diff
@@ -210,7 +211,9 @@ class Serializer @Inject() (mailboxIdFactory: MailboxId.Factory) {
   private implicit def notFoundWrites(implicit mailboxIdWrites: Writes[MailboxId]): Writes[NotFound] =
     notFound => JsArray(notFound.value.toList.map(mailboxIdWrites.writes))
 
-  private implicit val mailboxGetResponseWrites: Writes[MailboxGetResponse] = Json.writes[MailboxGetResponse]
+  private implicit def mailboxGetResponseWrites(implicit mailboxWrites: Writes[Mailbox]): Writes[MailboxGetResponse] = {
```
Comment (blackheaven) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:214]: extra brackets

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala (hunk 3 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala:54]: from the not review commit. But it is a method that merge two MailboxGetResults in one. 
By merging its found mailbox and notFound ids.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala (hunk 2 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala]: the `then` method of restassured is an assert in disguise. It's how we assert what we expect from the http response when using restassured

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala (hunk 4 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala]: We removed this commit it as its place in the protocol validation not in this PR.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala (hunk 5 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala]: to allow string interpolation, using $variableName

Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodCapabilitiesContract.scala]: by the way we removed this test from this PR it as its place in the protocol level one

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala [hunk 8]
```diff
@@ -525,6 +746,72 @@ trait MailboxGetMethodContract {
...
+               |    "urn:ietf:params:jmap:mail"],
+               |  "methodCalls": [[
+               |    "Mailbox/get",
+               |    {
+               |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+               |      "ids": ["${mailboxId}"]
+               |    },
+               |    "c1"]]
+               |}""".stripMargin)
+      .when
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala:769]: indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala [hunk 9]
```diff
@@ -525,6 +746,72 @@ trait MailboxGetMethodContract {
...
+      .post
+      .`then`
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxGetMethodContract.scala:771]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala [hunk 10]
```diff
@@ -65,22 +66,35 @@ class MailboxGetMethod @Inject() (serializer: Serializer,
             notFound = mailboxes.notFound))
           .map(mailboxGetResponse => Invocation(
             methodName = methodName,
-            arguments = Arguments(serializer.serialize(mailboxGetResponse).as[JsObject]),
+            arguments = Arguments(serializer.serialize(mailboxGetResponse)(writesWithFilteredProperties(capabilities)).as[JsObject]),
             methodCallId = invocation.methodCallId))))
   }
 
+  private def writesWithFilteredProperties(capabilities: Set[CapabilityIdentifier]): Writes[Mailbox] = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MailboxGetMethod.scala:73]: Shouldn't that method be in the `Serializer` instead?

**Comment by chibenwa** [comment]
`Caused by: java.io.FileNotFoundException: dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/blob.properties (No such file or directory)`

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**